lib/parser.ex

```diff
@@ -0,0 +1,205 @@
+        end
+
+      {:ok, data, packets, continued_packets}
+    end
+  end
+
+  defp verify_length_and_crc(data, segment_table) do
+    segments_count = Enum.count(segment_table)
+    content_length = Enum.sum(segment_table)
+
+    if 22 + 4 + 1 + segments_count + content_length > byte_size(data) do
+      {:error, :need_more_bytes}
+    else
+      <<before_crc::binary-size(22), crc::little-unsigned-size(32),
+        after_crc::binary-size(1 + segments_count + content_length), _rest::binary>> = data
+
+      crc_payload = <<before_crc::binary, 0::size(32), after_crc::binary>>
+
+      calculated_crc =
+        CRC.crc(
+          %{
+            extend: :crc_32,
+            init: 0x0,
+            poly: 0x04C11DB7,
+            xorout: 0x0,
+            refin: false,
+            refout: false
+          },
+          crc_payload
+        )
+
+      if calculated_crc == crc do
+        :ok
+      else
+        raise "Corrupted stream: invalid crc"
+      end
+    end
+  end
+
+  @spec prepend_continued_packet(
+          continued_packets :: %{},
+          bitstream_serial_number :: integer,
+          packets :: list(binary),
+          page_continued_packet :: binary | nil
+        ) ::
+          {packets :: list(binary), continued_packets :: %{},
+           page_continued_packet :: binary | nil}
+  defp prepend_continued_packet(
+         continued_packets,
+         bitstream_serial_number,
+         packets,
+         page_continued_packet
+       ) do
+    if Map.has_key?(continued_packets, bitstream_serial_number) do
+      # a special case when there is just one unfinished packet in the page
+      if Enum.empty?(packets) do
+        page_continued_packet =
+          continued_packets[bitstream_serial_number] <> page_continued_packet
+
+        continued_packets =
+          Map.put(continued_packets, bitstream_serial_number, page_continued_packet)
+
+        {packets, continued_packets, page_continued_packet}
+      else
+        [first_packet | rest_of_packets] = packets
+
+        packets = [
+          continued_packets[bitstream_serial_number] <> first_packet | rest_of_packets
+        ]
+
+        continued_packets = Map.delete(continued_packets, bitstream_serial_number)
+        {packets, continued_packets, page_continued_packet}
+      end
+    else
+      {packets, continued_packets, page_continued_packet}
+    end
+  end
+
+  @spec parse_header(data :: binary) ::
+          {:ok, data :: binary, header_type :: integer, bitstream_serial_number :: integer,
+           number_of_page_segments :: integer}
+          | {:error, :need_more_bytes}
+  defp parse_header(
+         <<"OggS", 0, header_type::unsigned-size(8), _granule_position::unsigned-size(64),
+           bitstream_serial_number::little-unsigned-size(32),
+           _page_sequence_number::unsigned-size(32), _crc::unsigned-size(32),
+           number_of_page_segments::unsigned-size(8), rest::binary>>
```

Comment:
The `unsigned` is used by default and when you omit the number it is one byte, so we can write:
```suggestion
         <<"OggS", 0, header_type, _granule_position::64, bitstream_serial_number::little-unsigned-size(32),
           _page_sequence_number::32, _crc::32, number_of_page_segments, rest::binary>>
```           

You can read more in the `iex` when you type `h <<>>`